## 4.8 Grouping Data & Aggregating Variables

### This script will contain the following:

#### Importing Libraries
#### Importing Orders Products Merged Dataframe
#### Group By & Aggregating
#### Creating a Loyalty Flag
#### Checking Spending Habits
#### Creating a Spending Flag
#### Creating a Frequent User Flag
#### Remove Unwanted Data
#### Exporting Dataframe

# 01. Importing Libraries

In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os

# 02. Importing Orders Products Merged Dataframe

In [2]:
# Path Creation
path = r'/Users/tyrasmussen/Desktop/05-2023 Instacart Basket Analysis'

In [3]:
# Import Orders Products Merged Dataframe
ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'ords_prods_merge8ExerciseColumns.pkl'))

In [4]:
ords_prods_merge.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   3139998      138            28                   6                 11   
1   1977647      138            30                   6                 17   
2    389851      709             2                   0                 21   
3    652770      764             1                   3                 13   
4   1813452      764             3                   4                 17   

   ...    price_range_loc    busiest_days   busiest_hours  \
0  ...  Mid-range product  Regularly busy  Average orders   
1  ...  Mid-range product  Regularly busy  Average orders   
2  ...  Mid-range product    Busiest days  Average orders   
3  ...  Mid-range product  Regularly busy  Average orders   
4  ...  Mid-range product  Regularly busy  Average orders   

  busiest_period_of_day max_order      loyalty_flag mean_prices spending_flag  \
0        Average orders        32  Regular customer   14.790541  High spender   
1        Average orders        32  Regular customer   14.790541  High spender   
2        Average orders         5      New customer    3.062500   Low spender   
3        Average orders         3      New customer    2.017241   Low spender   
4        Average orders         3      New customer    2.017241   Low spender   

  median_days_since_prior_order  frequent_user_flag  
0                           8.0   Frequent customer  
1                           8.0   Frequent customer  
2                           8.0   Frequent customer  
3                           9.0   Frequent customer  
4                           9.0   Frequent customer  

[5 rows x 25 columns]

In [5]:
ords_prods_merge.shape

(32404859, 25)

# Group By & Aggregating

In [7]:
# Find the mean of 'order_number' grouped by 'department_id'
ords_prods_merge.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
1                15.457838
2                17.277920
3                17.170395
4                17.811403
5                15.215751
6                16.439806
7                17.225802
8                15.340650
9                15.895474
10               20.197148
11               16.170638
12               15.887671
13               16.583536
14               16.773669
15               16.165037
16               17.665606
17               15.694469
18               19.310397
19               17.177343
20               16.473447
21               22.902379

### There are many differences between the results, for example department_id 21 had a mean of 22.902 in the entire dataframe and 25.535 in the subset of the dataframe.

## Creating a Loyalty Flag

In [8]:
# Create 'max_order' column
ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)

In [9]:
ords_prods_merge.head(15)

Unnamed: 0_x  product_id                product_name  aisle_id  \
0              0           1  Chocolate Sandwich Cookies        61   
1              0           1  Chocolate Sandwich Cookies        61   
2              0           1  Chocolate Sandwich Cookies        61   
3              0           1  Chocolate Sandwich Cookies        61   
4              0           1  Chocolate Sandwich Cookies        61   
5              0           1  Chocolate Sandwich Cookies        61   
6              0           1  Chocolate Sandwich Cookies        61   
7              0           1  Chocolate Sandwich Cookies        61   
8              0           1  Chocolate Sandwich Cookies        61   
9              0           1  Chocolate Sandwich Cookies        61   
10             0           1  Chocolate Sandwich Cookies        61   
11             0           1  Chocolate Sandwich Cookies        61   
12             0           1  Chocolate Sandwich Cookies        61   
13             0           1  Chocolate Sandwich Cookies        61   
14             0           1  Chocolate Sandwich Cookies        61   

    department_id  prices  Unnamed: 0.1  Unnamed: 0_y  order_id  user_id  ...  \
0              19     5.8          1987          1987   3139998      138  ...   
1              19     5.8          1989          1989   1977647      138  ...   
2              19     5.8         11433         11433    389851      709  ...   
3              19     5.8         12198         12198    652770      764  ...   
4              19     5.8         12200         12200   1813452      764  ...   
5              19     5.8         12372         12372   1701441      777  ...   
6              19     5.8         13096         13096   1871483      825  ...   
7              19     5.8         14603         14603   1290456      910  ...   
8              19     5.8         17065         17065    369558     1052  ...   
9              19     5.8         17070         17070    589712     1052  ...   
10             19     5.8         22082         22082   1798115     1379  ...   
11             19     5.8         23713         23713    497193     1480  ...   
12             19     5.8         23715         23715   1727826     1480  ...   
13             19     5.8         23716         23716   2694698     1480  ...   
14             19     5.8         23937         23937   3171753     1494  ...   

    add_to_cart_order  reordered  _merge     busiest_day    price_range_loc  \
0                   5          0    both  Regularly busy  Mid-range product   
1                   1          1    both  Regularly busy  Mid-range product   
2                  20          0    both     Busiest day  Mid-range product   
3                  10          0    both  Regularly busy  Mid-range product   
4                  11          1    both      Least busy  Mid-range product   
5                   7          0    both  Regularly busy  Mid-range product   
6                   2          0    both  Regularly busy  Mid-range product   
7                   1          0    both  Regularly busy  Mid-range product   
8                   1          0    both  Regularly busy  Mid-range product   
9                   2          1    both  Regularly busy  Mid-range product   
10                  1          0    both  Regularly busy  Mid-range product   
11                  1          0    both  Regularly busy  Mid-range product   
12                  3          1    both      Least busy  Mid-range product   
13                  4          1    both  Regularly busy  Mid-range product   
14                  8          0    both  Regularly busy  Mid-range product   

      busiest_days   busiest_hours busiest_period_of_day max_order  \
0   Regularly busy  Average orders        Average orders        32   
1   Regularly busy  Average orders        Average orders        32   
2     Busiest days  Average orders        Average orders         5   
3   Regularly busy  Average orders       

In [10]:
# Creating a loyalty flag
ords_prods_merge.loc[ords_prods_merge['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'

In [11]:
ords_prods_merge.loc[(ords_prods_merge['max_order'] <= 40) & (ords_prods_merge['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'

In [12]:
ords_prods_merge.loc[ords_prods_merge['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [13]:
# Check new 'loyalty_flag' columns
ords_prods_merge['loyalty_flag'].value_counts(dropna = False)

Regular customer    15876776
Loyal customer      10284093
New customer         6243990
Name: loyalty_flag, dtype: int64

## Checking Spending Habits

In [14]:
ords_prods_merge.groupby('loyalty_flag').agg({'prices': ['mean', 'min', 'max']})

prices              
                       mean  min      max
loyalty_flag                             
Loyal customer    10.386336  1.0  99999.0
New customer      13.294670  1.0  99999.0
Regular customer  12.495717  1.0  99999.0

In [15]:
ords_prods_merge.columns

Index(['Unnamed: 0_x', 'product_id', 'product_name', 'aisle_id',
       'department_id', 'prices', 'Unnamed: 0.1', 'Unnamed: 0_y', 'order_id',
       'user_id', 'order_number', 'orders_day_of_week', 'order_hour_of_day',
       'days_since_prior_order', 'add_to_cart_order', 'reordered', '_merge',
       'busiest_day', 'price_range_loc', 'busiest_days', 'busiest_hours',
       'busiest_period_of_day', 'max_order', 'loyalty_flag'],
      dtype='object')

### The min and max prices of the products purchased are the same for our 3 loyalty groups. However, the mean ranges as the highest group is 'New customer', then 'Regular customer', and the lowest group is 'Loyal customer'.

## Creating a Spending Flag

In [16]:
# Create 'mean_prices' column
ords_prods_merge['mean_prices'] = ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.mean)

In [17]:
ords_prods_merge.head(10)

Unnamed: 0_x  product_id                product_name  aisle_id  \
0             0           1  Chocolate Sandwich Cookies        61   
1             0           1  Chocolate Sandwich Cookies        61   
2             0           1  Chocolate Sandwich Cookies        61   
3             0           1  Chocolate Sandwich Cookies        61   
4             0           1  Chocolate Sandwich Cookies        61   
5             0           1  Chocolate Sandwich Cookies        61   
6             0           1  Chocolate Sandwich Cookies        61   
7             0           1  Chocolate Sandwich Cookies        61   
8             0           1  Chocolate Sandwich Cookies        61   
9             0           1  Chocolate Sandwich Cookies        61   

   department_id  prices  Unnamed: 0.1  Unnamed: 0_y  order_id  user_id  ...  \
0             19     5.8          1987          1987   3139998      138  ...   
1             19     5.8          1989          1989   1977647      138  ...   
2             19     5.8         11433         11433    389851      709  ...   
3             19     5.8         12198         12198    652770      764  ...   
4             19     5.8         12200         12200   1813452      764  ...   
5             19     5.8         12372         12372   1701441      777  ...   
6             19     5.8         13096         13096   1871483      825  ...   
7             19     5.8         14603         14603   1290456      910  ...   
8             19     5.8         17065         17065    369558     1052  ...   
9             19     5.8         17070         17070    589712     1052  ...   

   reordered  _merge     busiest_day    price_range_loc    busiest_days  \
0          0    both  Regularly busy  Mid-range product  Regularly busy   
1          1    both  Regularly busy  Mid-range product  Regularly busy   
2          0    both     Busiest day  Mid-range product    Busiest days   
3          0    both  Regularly busy  Mid-range product  Regularly busy   
4          1    both      Least busy  Mid-range product  Regularly busy   
5          0    both  Regularly busy  Mid-range product  Regularly busy   
6          0    both  Regularly busy  Mid-range product  Regularly busy   
7          0    both  Regularly busy  Mid-range product  Regularly busy   
8          0    both  Regularly busy  Mid-range product  Regularly busy   
9          1    both  Regularly busy  Mid-range product  Regularly busy   

    busiest_hours busiest_period_of_day max_order      loyalty_flag  \
0  Average orders        Average orders        32  Regular customer   
1  Average orders        Average orders        32  Regular customer   
2  Average orders        Average orders         5      New customer   
3  Average orders        Average orders         3      New customer   
4  Average orders        Average orders         3      New customer   
5  Average orders        Average orders        26  Regular customer   
6  Average orders        Average orders         9      New customer   
7     Most orders           Most orders        12  Regular customer   
8  Average orders        Average orders        20  Regular customer   
9  Average orders        Average orders        20  Regular customer   

  mean_prices  
0   14.790541  
1   14.790541  
2    3.062500  
3    2.017241  
4    2.017241  
5   14.964602  
6    4.939394  
7    7.700000  
8   10.958333  
9   10.958333  

[10 rows x 25 columns]

In [18]:
# Create a Spending Flag
ords_prods_merge.loc[ords_prods_merge['mean_prices'] >= 10, 'spending_flag'] = 'High spender'

In [19]:
ords_prods_merge.loc[ords_prods_merge['mean_prices'] < 10, 'spending_flag'] = 'Low spender'

In [20]:
# Check new 'spending_flag' columns
ords_prods_merge['spending_flag'].value_counts(dropna = False)

High spender    20332281
Low spender     12072578
Name: spending_flag, dtype: int64

## Creating a Frequent User Flag

In [21]:
# Create 'median_days_since_prior_order' column
ords_prods_merge['median_days_since_prior_order'] = ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform(np.median)

In [22]:
ords_prods_merge.head(10)

Unnamed: 0_x  product_id                product_name  aisle_id  \
0             0           1  Chocolate Sandwich Cookies        61   
1             0           1  Chocolate Sandwich Cookies        61   
2             0           1  Chocolate Sandwich Cookies        61   
3             0           1  Chocolate Sandwich Cookies        61   
4             0           1  Chocolate Sandwich Cookies        61   
5             0           1  Chocolate Sandwich Cookies        61   
6             0           1  Chocolate Sandwich Cookies        61   
7             0           1  Chocolate Sandwich Cookies        61   
8             0           1  Chocolate Sandwich Cookies        61   
9             0           1  Chocolate Sandwich Cookies        61   

   department_id  prices  Unnamed: 0.1  Unnamed: 0_y  order_id  user_id  ...  \
0             19     5.8          1987          1987   3139998      138  ...   
1             19     5.8          1989          1989   1977647      138  ...   
2             19     5.8         11433         11433    389851      709  ...   
3             19     5.8         12198         12198    652770      764  ...   
4             19     5.8         12200         12200   1813452      764  ...   
5             19     5.8         12372         12372   1701441      777  ...   
6             19     5.8         13096         13096   1871483      825  ...   
7             19     5.8         14603         14603   1290456      910  ...   
8             19     5.8         17065         17065    369558     1052  ...   
9             19     5.8         17070         17070    589712     1052  ...   

      busiest_day    price_range_loc    busiest_days   busiest_hours  \
0  Regularly busy  Mid-range product  Regularly busy  Average orders   
1  Regularly busy  Mid-range product  Regularly busy  Average orders   
2     Busiest day  Mid-range product    Busiest days  Average orders   
3  Regularly busy  Mid-range product  Regularly busy  Average orders   
4      Least busy  Mid-range product  Regularly busy  Average orders   
5  Regularly busy  Mid-range product  Regularly busy  Average orders   
6  Regularly busy  Mid-range product  Regularly busy  Average orders   
7  Regularly busy  Mid-range product  Regularly busy     Most orders   
8  Regularly busy  Mid-range product  Regularly busy  Average orders   
9  Regularly busy  Mid-range product  Regularly busy  Average orders   

   busiest_period_of_day  max_order      loyalty_flag mean_prices  \
0         Average orders         32  Regular customer   14.790541   
1         Average orders         32  Regular customer   14.790541   
2         Average orders          5      New customer    3.062500   
3         Average orders          3      New customer    2.017241   
4         Average orders          3      New customer    2.017241   
5         Average orders         26  Regular customer   14.964602   
6         Average orders          9      New customer    4.939394   
7            Most orders         12  Regular customer    7.700000   
8         Average orders         20  Regular customer   10.958333   
9         Average orders         20  Regular customer   10.958333   

  spending_flag median_days_since_prior_order  
0  High spender                           8.0  
1  High spender                           8.0  
2   Low spender                           8.0  
3   Low spender                           9.0  
4   Low spender                           9.0  
5  High spender                          11.0  
6   Low spender                          20.0  
7   Low spender                           6.0  
8  High spender                          10.0  
9  High spender                          10.0  

[10 rows x 27 columns]

In [23]:
# Creating a frequent user flag
ords_prods_merge.loc[ords_prods_merge['median_days_since_prior_order'] > 20, 'frequent_user_flag'] = 'Non-frequent customer'

In [28]:
ords_prods_merge.loc[(ords_prods_merge['median_days_since_prior_order'] > 10) & (ords_prods_merge['median_days_since_prior_order'] <= 20), 'frequent_user_flag'] = 'Regular customer'

In [24]:
ords_prods_merge.loc[ords_prods_merge['median_days_since_prior_order'] <= 10, 'frequent_user_flag'] = 'Frequent customer'

In [29]:
# Check new 'frequent_user_flag' columns
ords_prods_merge['frequent_user_flag'].value_counts(dropna = False)

Frequent customer        21559853
Regular customer          7208564
Non-frequent customer     3636437
NaN                             5
Name: frequent_user_flag, dtype: int64

In [30]:
ords_prods_merge.head()

Unnamed: 0_x  product_id                product_name  aisle_id  \
0             0           1  Chocolate Sandwich Cookies        61   
1             0           1  Chocolate Sandwich Cookies        61   
2             0           1  Chocolate Sandwich Cookies        61   
3             0           1  Chocolate Sandwich Cookies        61   
4             0           1  Chocolate Sandwich Cookies        61   

   department_id  prices  Unnamed: 0.1  Unnamed: 0_y  order_id  user_id  ...  \
0             19     5.8          1987          1987   3139998      138  ...   
1             19     5.8          1989          1989   1977647      138  ...   
2             19     5.8         11433         11433    389851      709  ...   
3             19     5.8         12198         12198    652770      764  ...   
4             19     5.8         12200         12200   1813452      764  ...   

     price_range_loc    busiest_days   busiest_hours  busiest_period_of_day  \
0  Mid-range product  Regularly busy  Average orders         Average orders   
1  Mid-range product  Regularly busy  Average orders         Average orders   
2  Mid-range product    Busiest days  Average orders         Average orders   
3  Mid-range product  Regularly busy  Average orders         Average orders   
4  Mid-range product  Regularly busy  Average orders         Average orders   

   max_order      loyalty_flag mean_prices spending_flag  \
0         32  Regular customer   14.790541  High spender   
1         32  Regular customer   14.790541  High spender   
2          5      New customer    3.062500   Low spender   
3          3      New customer    2.017241   Low spender   
4          3      New customer    2.017241   Low spender   

  median_days_since_prior_order frequent_user_flag  
0                           8.0  Frequent customer  
1                           8.0  Frequent customer  
2                           8.0  Frequent customer  
3                           9.0  Frequent customer  
4                           9.0  Frequent customer  

[5 rows x 28 columns]

# Remove Unwanted Data

In [6]:
ords_prods_merge.columns

Index(['product_id', 'product_name', 'aisle_id', 'department_id', 'prices',
       'order_id', 'user_id', 'order_number', 'orders_day_of_week',
       'order_hour_of_day', 'days_since_prior_order', 'add_to_cart_order',
       'reordered', '_merge', 'busiest_day', 'price_range_loc', 'busiest_days',
       'busiest_hours', 'busiest_period_of_day', 'max_order', 'loyalty_flag',
       'mean_prices', 'spending_flag', 'median_days_since_prior_order',
       'frequent_user_flag'],
      dtype='object')

In [6]:
# Remove 'Unnamed: 0_x' column from ords_prods_merge dataframe
ords_prods_merge = ords_prods_merge.drop(columns = ['Unnamed: 0_x'])

In [7]:
# Remove 'Unnamed: 0.1' column from ords_prods_merge dataframe
ords_prods_merge = ords_prods_merge.drop(columns = ['Unnamed: 0.1'])

In [8]:
# Remove 'Unnamed: 0_y' column from ords_prods_merge dataframe
ords_prods_merge = ords_prods_merge.drop(columns = ['Unnamed: 0_y'])

In [7]:
# Check columns
ords_prods_merge.columns

Index(['product_id', 'product_name', 'aisle_id', 'department_id', 'prices',
       'order_id', 'user_id', 'order_number', 'orders_day_of_week',
       'order_hour_of_day', 'days_since_prior_order', 'add_to_cart_order',
       'reordered', '_merge', 'busiest_day', 'price_range_loc', 'busiest_days',
       'busiest_hours', 'busiest_period_of_day', 'max_order', 'loyalty_flag',
       'mean_prices', 'spending_flag', 'median_days_since_prior_order',
       'frequent_user_flag'],
      dtype='object')

In [8]:
ords_prods_merge.isnull().sum()

product_id                             0
product_name                           0
aisle_id                               0
department_id                          0
prices                                 0
order_id                               0
user_id                                0
order_number                           0
orders_day_of_week                     0
order_hour_of_day                      0
days_since_prior_order           2076096
add_to_cart_order                      0
reordered                              0
_merge                                 0
busiest_day                            0
price_range_loc                        0
busiest_days                           0
busiest_hours                          0
busiest_period_of_day                  0
max_order                              0
loyalty_flag                           0
mean_prices                            0
spending_flag                          0
median_days_since_prior_order          5
frequent_user_fl

In [9]:
ords_prods_merge.shape

(32404859, 25)

In [10]:
# Removing missing values
ords_prods_merge_clean = ords_prods_merge[ords_prods_merge['days_since_prior_order'].isnull() == False]

In [11]:
# Removing missing values
ords_prods_merge_clean = ords_prods_merge[ords_prods_merge['median_days_since_prior_order'].isnull() == False]

In [12]:
# Removing missing values
ords_prods_merge_clean = ords_prods_merge[ords_prods_merge['frequent_user_flag'].isnull() == False]

In [13]:
ords_prods_merge_clean.shape

(32404854, 25)

# Exporting Dataframe

In [14]:
# Export data to pkl
ords_prods_merge_clean.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'ords_prods_merge_clean_8.pkl'))